p2p/transport/quic/transport.go

```diff
@@ -38,17 +38,118 @@ var quicDialContext = quic.DialContext // so we can mock it in tests
 
 var HolePunchTimeout = 5 * time.Second
 
+const bufferCap = 5
+
+type bufferEntry struct {
+	handshakes, total uint
+}
+
+type status struct {
+	cycle, pos int
+	check      bool
+}
+
+type accept struct {
+	buffer    []bufferEntry
+	mtx       sync.Mutex
+	duration  time.Duration
+	fraction  float64
+	startTime time.Time
+	current   status
+}
+
+func newAccept(duration time.Duration, fraction float64) *accept {
+	a := new(accept)
+
+	if duration == 0 {
+		duration = 1 * time.Minute
+	}
+	if fraction == 0 {
+		fraction = 0.5
+	}
+
+	a.duration = duration
+	a.fraction = fraction
+	a.startTime = time.Now()
+	a.buffer = make([]bufferEntry, 1, bufferCap)
+	return a
+}
+
+// Call it from inside a mutex to avoid a data race
+func (a *accept) getPosition() (bool, int) {
+	currentCycle := int(time.Since(a.startTime) / a.duration)
+	if currentCycle > a.current.cycle {
+		if len(a.buffer) < cap(a.buffer) {
+			a.buffer = append(a.buffer, bufferEntry{})
+			a.current = status{
+				cycle: currentCycle,
+				pos:   len(a.buffer) - 1,
+				check: false,
+			}
+		} else {
+			a.current = status{
+				cycle: currentCycle,
+				pos:   a.current.cycle % len(a.buffer),
+				check: true,
+			}
+			a.buffer[a.current.pos] = bufferEntry{}
+		}
+	}
+	return a.current.check, a.current.pos
+}
+
+func (a *accept) newHandshake() {
+	a.mtx.Lock()
+	_, i := a.getPosition()
+	a.buffer[i].handshakes += 1
+	a.mtx.Unlock()
+}
+
+func (a *accept) acceptFunction(clientAddr net.Addr, token *quic.Token) bool {
+	success := true
+	a.mtx.Lock()
+	checkRatio, index := a.getPosition()
+	if token == nil {
+		if checkRatio {
+			sumAll := bufferEntry{}
+			for i := range a.buffer {
+				sumAll = bufferEntry{
+					handshakes: sumAll.handshakes + a.buffer[i].handshakes,
+					total:      sumAll.total + a.buffer[i].total,
+				}
+			}
+
+			// Check Retry fail ratio
+			if sumAll.total > 0 && float64(sumAll.handshakes)/float64(sumAll.total) < a.fraction {
+				success = false
+			}
+		}
+		if success {
```

Comment:
That's to prevent double counting on the total for an incoming connection. If the fraction exceeded the fail ratio, you then receive a token != nil.
With `if success` in token== nil case, we avoid to count when we expect to receive another packet with the token verify: without this every new connection will count in the total as 2, therefore risking not to reach the right ratio to return to normality.

For all the others proposed changes, we implemented the solution.